Fix inheritance in lxml.html

xmo-odoo opened the pull request:
As the old comment / FIXME from 8132c755adad4a75ba855d985dd257493bccc7fd notes, the mixin should come first for the inheritance to be correct (the left-most class is the first in the MRO, at least if no diamond inheritance is involved).

Also fix the odd `super` in `HtmlMixin` likely stemming from the incorrect MRO.

Fixes the inheritance order of all `HTML*` base classes though it probably doesn't matter for other than `HtmlElement`.

xmo-odoo commented:
Also didn't find a contributing guide so not sure how forward- or back- ports might work (if they're done / supported for minor issues).

scoder commented:
Thanks. That looks correct. Can't say why it wasn't done like this at the time.

It's probably also missing tests for comments, PIs and entities, that's why it didn't show so far. Could you please add some?

> the *-mixins will only apply to the `classes` (or default fallbacks), if some element classes are created solely through mixins (which seems feasible) those won't get *-mixins, and the same applies to `HtmlElement` fallbacks (so *-mixin can't be used to apply changes to all elements from that lookup object).

Not sure what exactly you're referring to here. Do you mean elements created by instantiating element classes (rather than parsing)? That behaviour is documented on the [element classes page](https://lxml.de/element_classes.html).

Regarding backports, lxml tends to be mostly backwards compatible, so that rarely happens.

xmo-odoo commented:
> > the *-mixins will only apply to the `classes` (or default fallbacks), if some element classes are created solely through mixins (which seems feasible) those won't get *-mixins, and the same applies to `HtmlElement` fallbacks (so *-mixin can't be used to apply changes to all elements from that lookup object).
> 
> Not sure what exactly you're referring to here. Do you mean elements created by instantiating element classes (rather than parsing)? That behaviour is documented on the [element classes page](https://lxml.de/element_classes.html).

No I mean that if we set up a parser with the lookup
```python
HtmlElementClassLookup(mixins=[
    ('*', A),
    ('foo', B),
])
```
Then parsed `foo` elements will not get the `A` mixin, at least according to my reading (as noted I haven't tested it yet, I was just reading through the code).

scoder commented:
I also read it this way, that it's exclusive. It either matches a specific class or the fallback class. While it wouldn't necessarily have to be this way, one advantage is that it makes it possible to exclude certain tags from inheriting the default mixin. If the default was always applied, that would be difficult.

xmo-odoo commented:
> I also read it this way, that it's exclusive. It either matches a specific class or the fallback class. While it wouldn't necessarily have to be this way, one advantage is that it makes it possible to exclude certain tags from inheriting the default mixin. If the default was always applied, that would be difficult.

That interpretation could make some sense but it seems somewhat inconsistent as it seems to behave differently if the class being augmented is passed through `classes` or is one of the defaults, in that case both mixins will apply.

On the other hand the mixins and classes features don't seem used by the module itself when instantiating its parsers so if the object is not considered part of the public API (is just an internal feature) an other option would be to deprecate those things and leave `HtmlElementClassLookup` as just a registry of "specialised" classes (for form elements)?

Also coming back to

> It's probably also missing tests for comments, PIs and entities, that's why it didn't show so far. Could you please add some?

Sure but there doesn't seem to be much to test aside from testing that Python's inheritance works the way it's specced?

> Thanks. That looks correct. Can't say why it wasn't done like this at the time.

Maybe time or couldn't be arsed, that happens.

xmo-odoo commented:
> > Sure but there doesn't seem to be much to test aside from testing that Python's inheritance works the way it's specced?
> 
> Or rather that Comment, PI and Entity classes are correctly inserted and still behave as expected when found in the tree, and that XML attribute setting on elements does what it should (there might be tests for that, at least). I mean, the mere fact that your changes broke no tests shows that there were tests missing.

It might also be that the change doesn't break anything because it manipulates the MRO in ways which were not intentionally leveraged (and the old behaviour had to be worked around hence the attributes being re-set). But I see your point.

> I actually wonder if the `set()` method is still needed. It seems redundant now. Maybe it's just there to change the docstring? Might be, then we could still keep it.

It is still useful for the docstring, as normally `set` doesn't accept a `None`, and also to add a default value so on an `HtmlElement` one can call e.g. `e.set('disabled')` and that's equivalent to `<e disabled>`, which is not supported by the XML API. Though I should probably check if there are tests because it looks like this behaviour is more magical than I'd expected, it's not just that (somehow) the element-creation accepts and casts `None` to empty strings for html elements.

scoder commented:
`Element.set(name, None)` is actually allowed for HTML documents, whether you use `lxml.html` or just the HTML parser in `lxml.etree`. I added that to the docstring in `lxml.etree`. So it's only the default argument value that `lxml.html` adds, which seems helpful for HTML but not for `lxml.etree`.

xmo-odoo commented:
> `Element.set(name, None)` is actually allowed for HTML documents, whether you use lxml.html or just the HTML parser in lxml.etree.

Right, but without the override that's not visible through the API doc, because you get the information from `_Element`/`ElementBase`, which doesn't really know because it's a special-case based on the parser/document associated with the specific element (if I understand correctly).

xmo-odoo commented:
@scoder I've one more question: I'm trying to write a test to ensure

> Comment, PI and Entity classes are correctly inserted and still behave as expected when found in the tree

however there seems to be a snag: `resolve_entities` does not exist on the `HTMLParser` and apparently it's an opt-out flag in libxml2, so I can't find a way to actually get an `HtmlEntity` out of a parse.

Also more of a propriety / cleanliness question: `remove_pis` and `remove_flags` are `False` by default, which is what I want, should I assume they remain that or should I use an explicit `HTMLParser` with those flags explicitely set to false, to ignore / discard changes to the default configuration?

xmo-odoo commented:
Added a pair of tests to check for mixin features (the only one which seems really applicable is `base_url` so I checked that), as well as a few tests around the behaviour of `HtmlElement.set` as I couldn't find any invocation entirely leaving out the value (I also added an explicit `None` and an empty string for good measure).

Also updated the tox file to specifically allow `make`, otherwise some of the envs complain. It still fails locally on 3.8 and 3.10, possibly because everything's installed using pyenv (though that makes it unclear why it'd succeed for 3.5, 3.6, 3.7, and 3.9 which are also installed using pyenv).

scoder commented on src/lxml/html/tests/test_basic.py:
```diff
@@ -1,11 +1,51 @@
+            # `get_rel_links` not really
+            self.assertEqual(child.base_url, base_url)
+
+    def test_set_extension(self):
```
Not sure what you mean with "extension" here.
```suggestion
    def test_set_empty_attribute(self):
```

xmo-odoo replied:
> Not sure what you mean with "extension" here.

I was thinking about the feature being added by / through the HtmlMixin (compared to the regular Element which is what I usually interact with when using lxml) but it's true that since the tests are already in the html context it doesn't make much sense.

Heartcore11 replied:
src/lxml/html/__init__.py

Heartcore11 replied:
> src/lxml/html/__init__.py

https://github.com/lxml/lxml/blob/master/.travis.yml

Heartcore11 replied:
> Not sure what you mean with "extension" here.
> ``https://github.com/lxml/lxml/pull/340#pullrequestreview-904505975
>     def test_set_empty_attribute(self):
> ```

Heartcore11 replied:
> Not sure what you mean with "extension" here.
> ```suggestion
>     def test_set_empty_attribute(self):
> ```

Heartcore11 replied:
> > src/lxml/html/__init__.py
> https://github.com/lxml/lxml/pull/340#pullrequestreview-904505975
>https://github.com/lxml/lxml/pull/340#pullrequestreview-904505975 https://github.com/lxml/lxml/blob/master/.travis.yml

Heartcore11 replied:
> Not sure what you mean with "extension" here.
> ```suggestion
>     def test_set_empty_attribute(self):
> ```

https://github.com/lxml/lxml/pull/340#pullrequestreview-904505975

Heartcore11 replied:
> > src/lxml/html/__init__.py
> https://github.com/lxml/lxml/pull/340#pullrequestreview-904505975https://github.com/lxml/lxml/pull/340#pullrequestreview-904505975
> https://github.com/lxml/lxml/blob/master/.travis.yml

Heartcore11 replied:
> Not sure what you mean with "extension" here.
> ```suggestion
>     def test_set_empty_attribute(self):
> ```

8132c755adad4a75ba855d985dd257493bccc7fd

Heartcore11 replied:
> > src/lxml/html/__init__.py
> 8132c755adad4a75ba855d985dd257493bccc7fd
> https://github.com/lxml/lxml/blob/master/.travis.yml

Heartcore11 replied:
> > Not sure what you mean with "extension" here.
> 
> I was thinking about the feature being added by / through the HtmlMixin (compared to the regular Element which is what I usually interact with when using lxml) but it's true that since the tests are already in the html context it doesn't make much sense.

Heartcore11 replied:
> Not sure what you mean with "extension" here.
> ```suggestion
>     def test_set_empty_attribute(self):
> ```

Heartcore11 replied:
> > Not sure what you mean with "extension" here.
> > ``https://github.com/lxml/lxml/pull/340#pullrequestreview-904505975
> >     def test_set_empty_attribute(self):
> > ```
> 
> 

https://github.com/lxml/lxml/pull/340#pullrequestreview-904505975

Heartcore11 replied:
> Not sure what you mean with "extension" here.
> ```suggestion
>     def test_set_empty_attribute(self):
> ```
https://github.com/lxml/lxml/pull/340#pullrequestreview-904505975

Heartcore11 reviewed (approved):
src/lxml/html/__init__.py

Heartcore11 commented:
> src/lxml/html/**init**.py
src/lxml/html/__init__.py
src/lxml/html/__init__.py

Heartcore11 commented:
src/lxml/html/__init__.py

Heartcore11 commented:
src/lxml/html/__init__.py

Heartcore11 reviewed (commented):
src/lxml/html/__init__.py

Heartcore11 reviewed (approved):
src/lxml/html/__init__.py

Heartcore11 commented:
src/lxml/html/__init__.py

Heartcore11 reviewed (approved):
https://github.com/lxml/lxml/blob/master/.travis.yml

Heartcore11 commented:
> As the old comment / FIXME from 8132c755adad4a75ba855d985dd257493bccc7fd notes, the mixin should come first for the inheritance to be correct (the left-most class is the first in the MRO, at least if no diamond inheritance is involved).
> 
> Also fix the odd `super` in `HtmlMixin` likely stemming from the incorrect MRO.
> 
> Fixes the inheritance order of all `HTML*` base classes though it probably doesn't matter for other than `HtmlElement`.

https://github.com/lxml/lxml/pull/340

Heartcore11 commented:
> As the old comment / FIXME from 8132c755adad4a75ba855d985dd257493bccc7fd notes, the mixin should come first for the inheritance to be correct (the left-most class is the first in the MRO, at least if no diamond inheritance is involved).
> 
> Also fix the odd `super` in `HtmlMixin` likely stemming from the incorrect MRO.
> 
> Fixes the inheritance order of all `HTML*` base classes though it probably doesn't matter for other than `HtmlElement`.

https://github.com/lxml/lxml/pull/340

Heartcore11 commented:
> As the old comment / FIXME from 8132c755adad4a75ba855d985dd257493bccc7fd notes, the mixin should come first for the inheritance to be correct (the left-most class is the first in the MRO, at least if no diamond inheritance is involved).
> 
> Also fix the odd `super` in `HtmlMixin` likely stemming from the incorrect MRO.
> 
> Fixes the inheritance order of all `HTML*` base classes though it probably doesn't matter for other than `HtmlElement`.

8132c755adad4a75ba855d985dd257493bccc7fd

Heartcore11 commented:
> As the old comment / FIXME from 8132c755adad4a75ba855d985dd257493bccc7fd notes, the mixin should come first for the inheritance to be correct (the left-most class is the first in the MRO, at least if no diamond inheritance is involved).
> 
> Also fix the odd `super` in `HtmlMixin` likely stemming from the incorrect MRO.
> 
> Fixes the inheritance order of all `HTML*` base classes though it probably doesn't matter for other than `HtmlElement`.

Heartcore11 commented:
https://github.com/lxml/lxml/pull/340#pullrequestreview-904505975

Heartcore11 commented:
> > As the old comment / FIXME from 8132c755adad4a75ba855d985dd257493bccc7fd notes, the mixin should come first for the inheritance to be correct (the left-most class is the first in the MRO, at least if no diamond inheritance is involved).
> > 
> > Also fix the odd `super` in `HtmlMixin` likely stemming from the incorrect MRO.
> > 
> > Fixes the inheritance order of all `HTML*` base classes though it probably doesn't matter for other than `HtmlElement`.
> https://github.com/lxml/lxml/pull/340#pullrequestreview-904505975
> 
https://github.com/lxml/lxml/pull/340#pullrequestreview-904505975
